cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,367 @@
+func upgrade(rootArgs *rootArgs, args *upgradeArgs, l *logger) (err error) {
+	initLogsOrExit(rootArgs)
+	l.logAndPrintf("Client - istioctl version: %s\n", opversion.OperatorVersionString)
+
+	// Generates values for args.inFilename ICP specs yaml
+	targetValues, err := genProfile(true, args.inFilename, "",
+		"", "", args.force, l)
+	if err != nil {
+		l.logAndPrintf("Abort. Failed to generate values from file: %v, error: %v", args.inFilename, err)
+		return err
+	}
+
+	// Generate ICPS objects
+	_, targetICPS, err := genICPS(args.inFilename, "", "", args.force, l)
+	if err != nil {
+		l.logAndPrintf("Failed to generate ICPS from file %s, error: %s",
+			args.inFilename, err)
+		return err
+	}
+
+	// Get the target version from the tag in the ICPS
+	targetVersion := targetICPS.GetTag()
+	l.logAndPrintf("Upgrade - target version: %s\n", targetVersion)
+
+	// Create a kube client from args.kubeConfigPath and  args.context
+	kubeClient, err := manifest.NewClient(args.kubeConfigPath, args.context)
+	if err != nil {
+		l.logAndPrintf("Abort. Failed to connect Kubernetes API server, error: %v", err)
+		return err
+	}
+
+	// Get Istio control plane namespace
+	//TODO(elfinhe): support components distributed in multiple namespaces
+	istioNamespace := targetICPS.GetDefaultNamespace()
+
+	// Read the current Istio version from the the cluster
+	currentVer, err := retrieveControlPlaneVersion(kubeClient, istioNamespace, l)
+	if err != nil {
+		l.logAndPrintf("Failed to read the current Istio version, error: %v", err)
+		if !args.force {
+			return err
+		}
+	}
+
+	// Read the current Istio installation values from the cluster
+	currentValues, err := readValuesFromInjectorConfigMap(kubeClient, istioNamespace)
+	if err != nil {
+		l.logAndPrintf("Failed to read the current Istio installation values, "+
+			"error: %v", err)
+		if !args.force {
+			return err
+		}
+	}
+
+	// Check if the upgrade currentVer -> targetVersion is supported
+	err = checkSupportedVersions(currentVer, targetVersion, args.versionsURI, l)
+	if err != nil {
+		l.logAndPrintf("Upgrade version check failed: %v -> %v. Error: %v",
+			currentVer, targetVersion, err)
+		if !args.force {
+			return err
+		}
+	} else {
+		l.logAndPrintf("Upgrade version check passed: %v -> %v.\n", currentVer, targetVersion)
+	}
+
+	checkUpgradeValues(currentValues, targetValues, l)
+	waitForConfirmation(args.yes, l)
+
+	// Run pre-upgrade hooks
+	// TODO(elfinhe): add err handling when hook refactoring is done
+	runPreUpgradeHooks(kubeClient, istioNamespace,
+		currentVer, targetVersion, currentValues, targetValues, rootArgs.dryRun, l)
+
+	// Apply the Istio Control Plane specs reading from inFilename to the cluster
+	err = genApplyManifests(nil, args.inFilename, rootArgs.dryRun,
+		rootArgs.verbose, args.kubeConfigPath, args.context, upgradeWaitSecWhenApply, l)
+	if err != nil {
+		l.logAndPrintf("Failed to apply the Istio Control Plane specs. Error: %v", err)
```

Comment:
since you're already returning error, you can just log at the caller and return fmt.Errorf("failed to apply the IstioControlPlane spec: %v", err")
here and throughout